### PR TITLE
Rename 'MetropolisSampler.n_sweeps' to 'sweep_size' #1657

### PR DESCRIPTION
**Pull request by iamoeg**
#### Changes:

This PR applies the following changes:

- Rename `MetropolisSampler.n_sweeps` to `sweep_size`
- Add deprecation warning when `n_sweeps` is used when constructing the sampler
- Update all code references to `n_sweeps` to use `sweep_size` instead
- Update documentation (docs, tutorials, and examples) to use `sweep_size`
- Adjust tests (sampler, variational, and models) to use `sweep_size`
- Update the Changelog to document the deprecation of `n_sweeps`

Use diff for file-level changes.

#### Test Results:

- Some tests failed or raised a warning, but most of them are relative to the `jax` package
- The other tests that failed or raised a warning are flagged as "flaky" in the test suite itself (e.g. `sampler.test_sampler.test_correct_sampling`)

See test results attached for details.

#### Attachments:

Text output files of test results for `sampler`, `variational`, `models`, as well as the full test suite:

- [test-results-full.txt](https://github.com/netket/netket/files/13742470/test-results-full.txt)
- [test-results-models.txt](https://github.com/netket/netket/files/13742471/test-results-models.txt)
- [test-results-sampler.txt](https://github.com/netket/netket/files/13742472/test-results-sampler.txt)
- [test-results-variational.txt](https://github.com/netket/netket/files/13742473/test-results-variational.txt)

#### Related Issues:

- #1657

### CODE REVIEW

**Review by PhilipVinc — COMMENTED** [review]
Thanks! Looks great!

Just a minor thing: can you add as a property `n_sweep` to `MetropolisSampler` which throws a deprecation warning and suggests to use `sweep_size`, to minimise breakage?

Thanks a lot!

**Comment by iamoeg** [comment]
> Just a minor thing: can you add as a property `n_sweep` to `MetropolisSampler` which throws a deprecation warning and suggests to use `sweep_size`, to minimise breakage?

Just to make sure I understand, you want `n_sweep` **in addition to** `n_sweeps` and `sweep_size`?

File: netket/sampler/metropolis.py [hunk 1]
```diff
@@ -314,8 +315,17 @@ def __init__(
...
-        if n_sweeps is None:
-            n_sweeps = hilbert.size
+        if n_sweeps is not None:
+            warn_deprecation(
+                "Specifying `n_sweeps` when constructing sampler is deprecated. Please use `sweep_size` instead."
+            )
+            if sweep_size is None:
+                sweep_size = n_sweeps
+            else:
+                raise ValueError("Cannot specify both `sweep_size` and `n_sweeps`")
```
[the remark this answers is not in the document]
Comment (iamoeg) [reply netket/sampler/metropolis.py]: Noted, it's been changed in the latest commit.

**Comment by iamoeg** [comment]
> Just a minor thing: can you add as a property `n_sweep` to `MetropolisSampler` which throws a deprecation warning and suggests to use `sweep_size`, to minimise breakage?

Of course! It's been added in the latest commit [c8f862f](https://github.com/netket/netket/pull/1680/commits/c8f862f36431b98edd06fb8526d66fbd39f59088)

> Thanks a lot!

You're welcome, I'm glad to help!

**Comment by PhilipVinc** [comment]
Hmm no what I meant was rather to add the following

```python
    @property
    def n_sweep(self):
          warn_deprecation(...)
          return self.sweep_size
```

**Comment by iamoeg** [comment]
> Hmm no what I meant was rather to add the following
> 
> ```python
>     @property
>     def n_sweeps(self):
>           warn_deprecation(...)
>           return self.sweep_size
> ```

Oh, you meant using the property decorator! Sorry about the misunderstanding. Should I add a setter as well?

**Comment by PhilipVinc** [comment]
No, those are frozen data classes so can't be modified after construction.

**Comment by iamoeg** [comment]
@PhilipVinc Did you get a chance to see the latest version? Let me know if there's anything else.

File: netket/sampler/metropolis.py [hunk 2]
```diff
@@ -326,9 +335,16 @@ def __init__(
...
         self.rule = rule
-        self.n_sweeps = n_sweeps
+        self.sweep_size = sweep_size
         self.n_chains = n_chains
 
+    @property
+    def n_sweeps(self):
+        warn_deprecation(
+            "Specifying `n_sweeps` when constructing sampler is deprecated. Please use `sweep_size` instead."
+        )
```
Comment (PhilipVinc) [inline netket/sampler/metropolis.py:345]:
```suggestion
            "`MetropolisSampler.n_sweeps` is deprecated. Please use `MetropolisSampler.sweep_size` instead."
        )
```

**Review by PhilipVinc — APPROVED** [review]
good thanks!

**Pull request merged**